src/script/lua_api/l_mapgen.cpp

```diff
@@ -1475,6 +1487,68 @@ int ModApiMapgen::l_generate_decorations(lua_State *L)
 }
 
 
+// generate_biomes(vm, p1, p2, [noise_filler_depth])
+int ModApiMapgen::l_generate_biomes(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+
+	EmergeManager *emerge = getServer(L)->getEmergeManager();
+	if (!emerge)
+		return 0;
+
+	Mapgen *mg_current = emerge->getCurrentMapgen();
+	if (!mg_current || !mg_current->biomegen ||
+			mg_current->biomegen->getType() != BIOMEGEN_ORIGINAL)
+		return 0;
+
+	const NodeDefManager *ndef = getServer(L)->getNodeDefManager();
+
+	MapgenBasic mg;
```

Comment:
Why couldn't you use `mg_current` for this? `MapgenBasic() = default;` does not initialize trivial data types which can quickly result in unpredictable/erroneous code.

How about this instead?
```C++
MapgenBasic *mg = dynamic_cast<MapgenBasic *>(mg_current);
FATAL_ERROR_IF(!mg, "Bug! This mapgen does not inherit from MapgenBasic.");
```
.. and make `MapgenSinglenode` inherit `MapgenBasic` like all other mapgens. If that does not work, you could still specify `virtual` functions in `Mapgen` which the inherited classes would overwrite.

Reply:
I thought redefining `MapgenSinglenode` was out of question, not sure why I was thinking that, maybe paramat was against it a year or so ago. I tried to stay consistent with what was done for `l_generate_ores` and `l_generate_decorations` but I see why it does not make sense to create a temporary object for biomes: it is a strictly mapgen-time method, that ores and decorations are not.

Will do that. `MapgenV6` should also inherit `MapgenBasic` in that case.

Reply:
Changing existing mapgens to inherit from a class they have no business with just for this does sounds like a bad idea to me.